header.go

```diff
@@ -0,0 +1,84 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package yarpc
+
+import "github.com/yarpc/yarpc-go/internal"
+
+// CanonicalizeHeaderKey canonicalizes the given header key for storage into
+// the Headers map.
+func CanonicalizeHeaderKey(k string) string {
+	return internal.CanonicalizeHeaderKey(k)
+}
+
+// Headers defines application headers which will be sent over the wire to the
+// recipient of an RPC call.
+type Headers internal.Headers
+
+// NewHeaders builds a new Headers object.
+func NewHeaders() Headers {
+	return Headers{}
+}
+
+// With returns a Headers object with the given key-value pair added to it.
+// If a header with the same name already exists, it will be overwritten.
+//
+// The returned Headers object MAY use a different underlying data store than
+// the original Headers object, and so the returned object MUST always be used
+// instead of the original object.
```

Comment:
maybe call out that this is similar to `append`? Don't want users thinking that `With` will return a new Headers always, since the old value may be modified and can't be reused